bevy_lint/src/lints/borrow_of_reborrowable.rs

```diff
@@ -0,0 +1,224 @@
+                    // `for<'a> (Commands<'_, '_>) -> EntityCommands<'a>`
+                    // without getting: `error[E0515]: cannot return value referencing function parameter `commands``
+                    continue;
+                }
+
+                let arg_ident = arg_names[arg_index];
+                let span = decl.inputs[arg_index].span.to(arg_ident.span);
+
+                span_lint_and_sugg(
+                    cx,
+                    reborrowable.lint(),
+                    span,
+                    reborrowable.message(),
+                    reborrowable.help(),
+                    reborrowable.suggest(arg_ident, peeled_ty.to_string()),
+                    // Not machine-applicable since the function body may need to
+                    // also be updated to account for the removed ref
+                    Applicability::MaybeIncorrect,
+                );
+            }
+        });
+    }
+}
+
+#[derive(Debug)]
+enum Reborrowable {
+    Commands,
+    // Deferred,
+    // DeferredWorld,
+    EntityCommands,
+    // EntityMut,
+    // EntityMutExcept,
+    // FilteredEntityMut,
+    // FilteredResourcesMut,
+    // Mut,
+    // MutUntyped,
+    NonSendMut,
+    // PtrMut,
+    Query,
+    // QueryIterationCursor,
+    ResMut,
+}
+
+impl Reborrowable {
+    fn try_from_ty<'tcx>(cx: &LateContext<'tcx>, ty: Ty<'tcx>) -> Option<Self> {
+        if match_type(cx, ty, &crate::paths::COMMANDS) {
+            Some(Self::Commands)
+        } else if match_type(cx, ty, &crate::paths::ENTITY_COMMANDS) {
+            Some(Self::EntityCommands)
+        } else if match_type(cx, ty, &crate::paths::QUERY) {
+            Some(Self::Query)
+        } else if match_type(cx, ty, &crate::paths::RES_MUT) {
+            Some(Self::ResMut)
+        } else if match_type(cx, ty, &crate::paths::NON_SEND_MUT) {
+            Some(Self::NonSendMut)
+        } else {
+            None
+        }
+    }
+
+    fn lint(&self) -> &'static Lint {
+        match self {
+            Self::Commands => BORROW_OF_COMMANDS.lint,
+            Self::EntityCommands => BORROW_OF_COMMANDS.lint,
+            Self::Query => BORROW_OF_QUERY.lint,
+            Self::ResMut => BORROW_OF_RESOURCE.lint,
+            Self::NonSendMut => BORROW_OF_RESOURCE.lint,
+        }
+    }
+
+    fn message(&self) -> String {
+        let name = self.name();
+        format!("parameter takes `&mut {name}` instead of a re-borrowed `{name}`",)
+    }
+
+    fn name(&self) -> &'static str {
+        match self {
+            Self::Commands => "Commands",
+            Self::EntityCommands => "EntityCommands",
+            Self::Query => "Query",
+            Self::ResMut => "ResMut",
+            Self::NonSendMut => "NonSendMut",
+        }
+    }
+
+    fn help(&self) -> String {
+        let name = self.name();
+        format!("use `{name}` instead")
+    }
+
+    fn suggest(&self, ident: Ident, ty: String) -> String {
+        format!("mut {ident}: {ty}")
```

Comment:
This prints out `mut commands: bevy::prelude::Commands<'_, '_>`, but it seems a bit silly for the type to be fully qualified and usually-invisible lifetimes to be present. Unfortunately, other types need this data (`mut _res: bevy::prelude::ResMut<'_, Data>`), so we can't replace it with a simple `format!("mut {ident}: {}", self.name())`.

Could you make a follow-up issue to improve the suggested output of this lint?

Reply:
Will do!